Prevent dropdown menu's scroll offset from going negative

mklim opened the pull request:
Before the offset was adjusted by `- (buttonTop - menuTop)`, which meant
that with no item selected the offset would actually be a negative
number. In long lists that would result in the the dropdown scrolling to
the very last item in the list. Changed to be the `selectedItemOffset`
directly instead, since the extra margin seemed unnecessary.

Fixes flutter/flutter#15346

mklim commented:
@HansMuller I dug through the history a bit but didn't see the reasoning behind `- (buttonTop - menuTop)`. From what I can tell it always computes a negative value, but maybe there's a case where that offset is needed that I'm not seeing? So far it looks like all the tests are passing with it removed still.

HansMuller commented:
@mklim - the overall idea is to have the menu appear with the selected item lined up with (itself) the dropdown menu button's value. Forgetting scrolling for the moment: if the menu contains 4 items [A, B, C, D] the selected item is C, the dropdown looks like "Hello: C v", and when the menu pops up it looks like:

```
      A
      B 
Hello C
      D
```

As you can see, the top of the menu is negative relative to the top of the dropdown button. 

Things get more complicated when it's also necessary to scroll the selected item into view of course.

mklim commented:
Thanks, makes sense. I'll add a test for that and rework this to make sure it still handles that case.

Commit:
Prevent dropdown menu offset from going negative

In long lists this resulted in the dropdown scrolling to the very last
item in its list. Now clamping the value at `0.0`. Added a test to
verify that the selected item aligns with the button to test the offset.

Fixes flutter/flutter#15346

mklim commented:
Updated.

HansMuller commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -346,9 +346,9 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
-      if (preferredMenuHeight > maxMenuHeight)
-        scrollOffset = selectedItemOffset - (buttonTop - menuTop);
+      final double scrollOffset = (preferredMenuHeight > maxMenuHeight) ?
+        (selectedItemOffset - (buttonTop - menuTop)).clamp(0.0, double.maxFinite)
```
This is OK, but it seems like an odd way to say: 
`math.max(0.0, selectedItemOffset - (buttonTop - menuTop))`

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
 
+  testWidgets('Dropdown menu scrolls to first item in long lists', (WidgetTester tester) async {
+    final Key buttonKey = UniqueKey();
+    Widget build() => buildFrame(
```
We typically only use `=>` for one-liners that return a value.

Why not just put the buildFrame() call in-line?

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+
+    // Open the dropdown menu
+    await tester.pumpWidget(build());
+    final RenderBox buttonBox = tester.renderObject(find.byKey(buttonKey));
```
`tester.renderObject<RenderBox>(...)`

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+    assert(buttonBox.attached);
+    await tester.tap(find.byKey(buttonKey));
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 10)); // finish the menu animation
```
`await tester.pumpAndSettle()`

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+    assert(firstItem.attached);
+
+    // List should be scrolled so that the first item is at the top
+    const Offset selectedItemOffset = Offset(0.0, -8.0);
```
The comment should explain what the -8.0 is for.

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+
+    // List should be scrolled so that the first item is at the top
+    const Offset selectedItemOffset = Offset(0.0, -8.0);
+    expect(
```
NICE.

mklim replied:
:)

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+    );
+
+    // Open the dropdown menu
+    await tester.pumpWidget(build());
```
As before. Typically we don't add a level of indirection like this.

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+    // Find the selected item in the scrollable dropdown list
+    final RenderBox selectedItem = tester.renderObject<RenderBox>(
+      find.descendant(of: find.byType(Scrollable), matching: find.byKey(const ValueKey<String>('50'))));
+    assert(selectedItem.attached);
```
Why are we asserting `attached` here and elsewhere? I don't believe WidgetTester.renderObject will succeed for an unattached renderer.

... I see, these assertions appear through this file, but not anywhere else (in fact I appear to responsible for a few of them).

HansMuller commented on packages/flutter/test/material/dropdown_test.dart:
```diff
@@ -422,6 +422,67 @@ void main() {
+    assert(buttonBox.attached);
+    await tester.tap(find.byKey(buttonKey));
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1)); // finish the menu animation
```
`await tester.pumpAndSettle()`

HansMuller reviewed (commented):
This all looks good. I'd just like to get to the bottom of the attached assertions. Maybe we're trying to determine if tester.renderObject is turning up scrollable child that's no longer visible? Seems like that should be checked with expect, not assert(). And not by checking the attached property, but by checking that it's within the scrollable's viewport.

Commit:
Review fixup.

mklim commented:
Thanks for the review! This patch looks much cleaner now.

> I'd just like to get to the bottom of the attached assertions.

They definitely weren't necessary for the test cases I added, I was just following the pattern I saw in other cases in this file. I assumed that it was common or best practice to assert that elements exist before acting on them since I saw it in other test cases. For my cases the last `expect()`s are checking everything that I need to actually verify. Not sure why the other test cases had them originally though, that is probably worth following up on.

HansMuller reviewed (approved):
LGTM

Eventually we should get to the bottom of the attached assertions in the rest of dropdown_test.dart. This PR looks good without them :-).

gspencergoog commented:
@fkorotkov, why is there still one GitHub check status that says it hasn't completed, when Cirrus says it has?  It looks like GitHub is listing `tests-windows` status three times in this PR.

fkorotkov commented:
@gspencergoog there are two of them for the same task! One passed and one reported as still queued Looking into it.

mklim commented:
@fkorotkov Thanks! I just recently tried restarting the task, if that makes a difference. The new run from that is https://cirrus-ci.com/task/5629948521873408.

Hixie commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -346,9 +346,8 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
-      if (preferredMenuHeight > maxMenuHeight)
-        scrollOffset = selectedItemOffset - (buttonTop - menuTop);
+      final double scrollOffset = (preferredMenuHeight > maxMenuHeight) ?
+        math.max(0.0, selectedItemOffset - (buttonTop - menuTop)) : 0.0;
```
what i find weird about this approach is that it's asymmetrical. Why do we clamp at zero but not at some equivalent large offset?

mklim replied:
I never saw a situation where `selectedItemOffset - (buttonTop - menuTop)` produced values that were incorrectly large, so I didn't see a need for it.

In my testing it looked like this did correctly calculate an offset for all items after the first one. It perfectly aligned them with the dropdown button itself if there was room to scroll to them. But it was consistently setting a negative value for when either nothing or the first item in a list was selected. In short lists that doesn't matter since there's no room to scroll, but as soon as there was a scrollbar then it was wrapping around to the bottom of the list to match the negative `scrollOffset`.

@jslavitz's fix in flutter/flutter#22594 looks more robust to me. If the problem is with the variables in this calculation, updating that makes more sense than clamping off known incorrect bounds.

jslavitz commented:
Hey! I've got a fix here: https://github.com/flutter/flutter/pull/22594 that I think makes sense.
